content/livesync/models.textile

```diff
@@ -0,0 +1,357 @@
+
+h2(#optimistic-updates). Add Optimistic Updates
+
+The Models SDK supports optimistic updates, a feature that enables you to immediately render changes to your data model before the backend confirms the changes, making updates appear instantaneous.
+
+Optimistic events are used to make local, provisional changes to your data, anticipating that your backend will eventually confirm or reject these changes. The optimistic event will be processed by your "merge function":#merge and included in the local model state optimistically.
+
+Use the @model.optimistic()@ method to apply an optimistic update to your model.
+
+1. Call @model.optimistic()@ on your model with the optimistic event.
+2. Apply the corresponding change to your backend.
+3. Await the confirmation of the optimistic update from the backend, or optionally cancel the optimistic update.
+
+The @model.optimistic()@ function returns a promise resolving to two values:
+
+* A confirmation promise that resolves when the backend confirms or rejects the optimistic update.
+* A function to explicitly cancel the optimistic update.
+
+The following demonstrates an optimistic update to implementing changes in the model:
+
+```[javascript]
+// optimistically apply the changes to the model
+const [confirmation, cancel] = await model.optimistic({
+    mutationId: 'my-mutation-id',
+    name: 'addComment',
+    data: 'New comment!',
+});
+
+try {
+    // apply the changes in your backend
+    await updatePost('my-mutation-id', 'New comment!');
+    // wait for the optimistic event to be confirmed
+    await confirmation;
+} catch (err) {
+    // something went wrong, cancel the optimistic update
+    cancel();
+}
+```
+
+<aside data-type='important'>
+<p>Be aware that updating the model state in your database happens entirely through your backend endpoint.</p>
+</aside>
+
+h3(#mutation-id). Mutation ID
+
+The SDK requires a mechanism to identify when a change event received from the backend matches an optimistic event that has already been applied locally.
+
+To achieve this, your clients need to assign a unique @mutationId@ to each optimistic event. This ID can be any string, though it's commonly a "UUID":https://www.npmjs.com/package/uuid. The @mutationId@ must be sent to your backend and included in the confirmation event that your backend writes to the outbox:
+
+```[sql]
+BEGIN;
+-- mutate your data, e.g.:
+INSERT INTO comments (comment) VALUES ('New comment!');
+-- write change event to outbox, e.g.:
+INSERT INTO outbox (mutation_id, channel, name, ...) VALUES ('my-mutation-id', 'posts:123', 'addComment', ...);
+COMMIT;
+```
+
+h2(#subscribe). Subscribe to changes in the model
+
+To receive realtime notifications when there are changes in the state, you can subscribe to changes to the changes.
```

Comment:
> ... you can subscribe to the model

The subscribe callback will be passed the updated model state when it changes. This is useful for updating elements in your application, for example the UI.